wkcuber/api/Layer.py

```diff
@@ -88,6 +105,74 @@ def set_bounding_box_size(self, size: Tuple[int, int, int]):
         for _, mag in self.mags.items():
             mag.view.size = size
 
+    def setup_mag(self, mag: Mag) -> None:
+        raise NotImplemented
+
+    def _initialize_mag_from_other_mag(self, new_mag_name, other_mag, compress):
+        raise NotImplemented
+
+    def downsample(
+            self,
+            from_mag: Mag,
+            max_mag: Mag,
+            interpolation_mode: str,
+            compress: bool,
+            scale: Tuple[float, float, float] = None,
+            buffer_edge_len=None,
+            args: Namespace = None,
+    ):
+        # if 'scale' is set, the data gets downsampled anisotropic
+        interpolation_mode = parse_interpolation_mode(interpolation_mode, self.name)
+        prev_mag = from_mag
+        target_mag = get_next_mag(prev_mag, scale)
+        while target_mag <= max_mag:
+            assert prev_mag < target_mag
+            assert target_mag.to_layer_name() not in self.mags
+
+            prev_mag_ds = self.mags[prev_mag.to_layer_name()]
+
+            mag_factors = [
+                t // s for (t, s) in zip(target_mag.to_array(), prev_mag.to_array())
+            ]
+
+            # initialize empty target mag
+            target_mag_ds = self._initialize_mag_from_other_mag(target_mag, prev_mag_ds, compress)
+            target_mag_view = target_mag_ds.get_view()
+
+            # perform downsampling
+            with get_executor_for_args(args) as executor:
+                voxel_count_per_cube = np.prod(prev_mag_ds._get_file_dimensions())
+                job_count_per_log = math.ceil(1024 ** 3 / voxel_count_per_cube)  # log every gigavoxel of processed data
+
+                num_chunks_per_dim = [math.ceil(s1/s2) for s1, s2 in zip(target_mag_view.size, target_mag_ds._get_file_dimensions())]
+
+                if buffer_edge_len is None:
+                    buffer_edge_len = determine_buffer_edge_len(prev_mag_ds.view) # DEFAULT_EDGE_LEN
+                job_args = (
+                    prev_mag_ds.get_view(),  # the offset and size of the view need to be set from inside the job
+                    mag_factors,
+                    interpolation_mode,
+                    buffer_edge_len,
+                    compress,
+                    target_mag_ds._get_file_dimensions(),
+                    job_count_per_log,
+                    target_mag_view.global_offset,
+                    num_chunks_per_dim,
+                    use_logging,
```

Comment:
In the old implementation, an counter `i` was used to determine which chunk should use logging. I am now using `.for_each_chunk(...)`, which passes the same arguments to all chunks. This is my attempt to get logging to work: a function `use_logging` is passed with which each chunks can calculate the counter `i` artificially and therefore know whether it should log or not. However the executor requires this function to be static (I think). Therefore I also have to pass a few additional variables to each chunk, so that they can call this function in a static way. This is a bit ugly.
What would be a better way to do this?